apps/lib/apps.c

```diff
@@ -3275,23 +3275,57 @@ void corrupt_signature(const ASN1_STRING *signature)
     s[signature->length - 1] ^= 0x1;
 }
 
+int check_cert_time_string(const char *time, const char *desc)
+{
+    if (time == NULL || strcmp(time, "today") == 0
+            || ASN1_TIME_set_string_X509(NULL, time))
+        return 1;
+    BIO_printf(bio_err,
+            "%s is invalid, it should be \"today\" or have format [CC]YYMMDDHHMMSSZ\n",
+            desc);
+    return 0;
+}
+
 int set_cert_times(X509 *x, const char *startdate, const char *enddate,
-                   int days)
+                   int days, int strict_compare_times)
 {
+    if (!check_cert_time_string(startdate, "start date"))
+        return 0;
+    if (!check_cert_time_string(enddate, "end date"))
+        return 0;
+    if (enddate != NULL && days != 0) {
+        BIO_printf(bio_err, "Warning: -not_after option overriding -days option\n");
```

Comment:
Just realised that for generality we should s/-not_after/-not_after or -enddate/

Reply:
Unfortunately, `-enddate` has different meaning for `ca.c` and `x509.c`: The former is just an alias, the latter causes print of the `notAfter` field. `req.c` doesn't know `-enddate` at all.
Is this be acceptable?
At least `-not_after` is unambiguous :grin:

Reply:
Looks like I should better keep this check in the respective app.

Reply:
Sigh, all those inconsistencies in the app options.
You can use this workaround: 
```
BIO_printf(bio_err, "Warning: -not_after%s option overriding -days option\n",
           strict_compare_times ? "" : " or -enddate");
```

Reply:
Doing this warning in each of the apps is more redundant.
If you do not like the workaround I suggested above, it could be factored out cleanly into `set_cert_times()` using a further parameter, or by changing the `strict_compare_times` parameter to e.g., `ca`.

Reply:
I'll give your workaround a try, hopefully tomorrow.

Reply:
Good. 
BTW, please also reduce (by suitable partial squashing) the large number of commits in this PR.

Reply:
> Good. BTW, please also reduce (by suitable partial squashing) the large number of commits in this PR.

I condensed everything to a single commit.

Reply:
> Doing this warning in each of the apps is more redundant. If you do not like the workaround I suggested above, the warning could be factored out cleanly into `set_cert_times()` using a further parameter, or by changing the `strict_compare_times` parameter to e.g., `ca`.

I see more problems in adding those checks to the lib than reduction of redundancy helps in this case.

The problem lies in the detection of intended setting of the `-days` option, which can be `!= 0` in case of `ca` or `!= DAYS_UNSET` otherwise. Properly handling this would make the library function more complex.

There is still a possibility to explicitly set the values to /unset/ without detecting this fact.
In the past I was used to add boolean variables that get set, whenever the corresponding option is used. Furthermore, this can also track double assignment of values.

Are you fine with leaving this as-is?

Reply:
There is a simple way to solve the problem and also to improve uniformity : 
use in `ca.c` the same logic as with `req.c`  and `x509.c`, initialising `days = DAYS_UNSET`.

Reply:
I do not think this will work unless changing the whole logic in `ca.c`: The `DAYS_UNSET = -1` only works for `req.c` and `x509.c` as the period must be positive. `ca.c` also allows for negative periods, hence `-1` is a valid argument to `-days`.
This cannot be distinguished from the `-days` unset like in the other two apps.

Reply:
I think we agreed on keeping the implementation as-is now.